narwhal/primary/src/core.rs

```diff
@@ -182,328 +180,308 @@ impl Core {
+        let mut requests: FuturesUnordered<_> = peers
+            .map(|peer| {
+                let header = header.clone();
+                Self::request_vote(
+                    network.clone(),
+                    committee.clone(),
+                    certificate_store.clone(),
+                    peer,
+                    header,
+                )
+            })
+            .collect();
+        loop {
+            if certificate.is_some() {
+                break;
             }
-            if header.round == round_digest_pair.round {
-                // check the hash first
-                let temp_vote = Vote::new(header, &self.name, &mut self.signature_service).await;
-                if temp_vote.digest() != round_digest_pair.vote_digest {
-                    // we already sent a vote for a different header to the authority for this round
-                    // don't equivocate by sending a different vote for the same round
-                    warn!(
-                        "Authority {} submitted duplicate header for votes at round {}",
-                        header.author, header.round
-                    );
-                    self.metrics
-                        .votes_dropped_equivocation_protection
-                        .with_label_values(&[&header.epoch.to_string(), header_source])
-                        .inc();
-                    return Ok(());
-                }
+            tokio::select! {
+                result = &mut requests.next() => {
+                    match result {
+                        Some(Ok(vote)) => {
+                            certificate = votes_aggregator.append(
+                                vote,
+                                &committee,
+                                &header,
+                            )?;
+                        },
+                        Some(Err(e)) => debug!("failed to get vote for header {header}: {e:?}"),
+                        None => break,
+                    }
+                },
+                _ = &mut cancel => {
+                    debug!("canceling Header proposal {header} for round {}", header.round);
+                    return Err(DagError::Canceled)
+                },
             }
         }
-        self.send_vote(header).await
-    }
 
-    #[instrument(level = "debug", skip_all)]
-    async fn send_vote(&mut self, header: &Header) -> DagResult<()> {
-        // Make a vote and send it to the header's creator.
-        let vote = Vote::new(header, &self.name, &mut self.signature_service).await;
-        debug!(
-            "Created vote {vote:?} for {header} at round {}",
-            header.round
-        );
-        let vote_digest = vote.digest();
+        // Check if we successfully formed a certificate.
+        let certificate =
+            certificate.ok_or_else(|| DagError::CouldNotFormCertificate(header.digest()))?;
+        debug!("Assembled {certificate:?}");
+        Ok(certificate)
```

Comment:
So my thinking was that it'd be better at the beginning of process_own_certificate because if some other stuff gunks up the Core loop before we call it then we'd probably want to include that latency as part of certificate creation latency. (Like a Zen koan...if we have created a certificate but never broadcast it, have we really created it?)

What are your thoughts on the tradeoffs - any reasons in particular it'd be better at the end of propose_header?

Reply:
That makes sense to me and we do make note that most of our latencies include the time taken for disseminating information to downstream channels. With that line of thinking the end of `process_certificate_internal` would complete the `header -> certificate` flow. At this time the certificate is [persisted in store ](https://github.com/MystenLabs/sui/blob/4b4a5a1828086f8197ca5156d0b01a3400e4dde0/narwhal/primary/src/core.rs#L550)and [sent to consensus](https://github.com/MystenLabs/sui/blob/4b4a5a1828086f8197ca5156d0b01a3400e4dde0/narwhal/primary/src/core.rs#L569) beginning the next phase. I think it makes sense to move the metrics to be updated then.